### PR TITLE
Enforce Content-Type requirement on the rest layer and remove deprecated methods

### PR DESCRIPTION
**Pull request by jaymode**
This commit enforces the requirement of Content-Type for the REST layer and removes the deprecated methods in transport requests and their usages.

While doing this, it turns out that there are many places where *Entity classes are used from the apache http client libraries and many of these usages did not specify the content type. The methods that do not specify a content type explicitly have been added to forbidden apis to prevent more of these from entering our code base.

Relates #19388

### CODE REVIEW

**Comment by clintongormley** [comment]
Please add breaking changes docs

**Comment by jaymode** [comment]
> Please add breaking changes docs

Thanks for the reminder. Breaking changes docs were added as part of #22691 and I cleaned them up some more.

File: buildSrc/src/main/resources/forbidden/es-test-signatures.txt [hunk 1]
```diff
@@ -25,3 +25,8 @@ org.apache.lucene.util.LuceneTestCase$Nightly @ We don't run nightly tests at th
 com.carrotsearch.randomizedtesting.annotations.Nightly @ We don't run nightly tests at this point!
 
 org.junit.Test @defaultMessage Just name your test method testFooBar
+
+@defaultMessage Explicitly specify the ContentType of HTTP entities
+org.apache.http.entity.StringEntity#<init>(java.lang.String)
+org.apache.http.entity.StringEntity#<init>(java.lang.String,java.lang.String)
+org.apache.http.entity.StringEntity#<init>(java.lang.String,java.nio.charset.Charset)
```
Comment (javanna) [inline buildSrc/src/main/resources/forbidden/es-test-signatures.txt]: should we also add ByteArrayEntity here? what is the difference between the http-signatures file below and this one?

File: client/rest/src/test/java/org/elasticsearch/client/RestClientSingleHostTests.java [hunk 2]
```diff
@@ -431,7 +432,7 @@ private HttpUriRequest performRandomRequest(String method) throws Exception {
         HttpEntity entity = null;
         boolean hasBody = request instanceof HttpEntityEnclosingRequest && getRandom().nextBoolean();
         if (hasBody) {
-            entity = new StringEntity(randomAsciiOfLengthBetween(10, 100));
+            entity = new StringEntity(randomAsciiOfLengthBetween(10, 100), ContentType.APPLICATION_JSON);
```
Comment (javanna) [inline client/rest/src/test/java/org/elasticsearch/client/RestClientSingleHostTests.java]: thanks for fixing all these.

File: core/src/main/java/org/elasticsearch/http/HttpTransportSettings.java [hunk 3]
```diff
@@ -69,7 +70,14 @@
...
     public static final Setting<Boolean> SETTING_HTTP_CONTENT_TYPE_REQUIRED =
-        Setting.boolSetting("http.content_type.required", false, Property.NodeScope);
+        new Setting<>("http.content_type.required", (s) -> Boolean.toString(true), (s) -> {
+            final boolean value = Booleans.parseBoolean(s);
+            if (value == false) {
+                throw new IllegalArgumentException("http.content_type.required cannot be set to false. It exists only to make a rolling" +
+                    " upgrade easier");
+            }
+            return true;
+        }, Property.NodeScope, Property.Deprecated);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/http/HttpTransportSettings.java]: do we have a test for this? just making sure that a 5.x cluster having the setting set to false can upgrade to 6.0.

File: docs/reference/api-conventions.asciidoc [hunk 4]
```diff
@@ -675,22 +675,15 @@ should also be passed with a media type value that indicates the format
 of the source, such as `application/json`.
 
 [float]
-=== Content-Type auto-detection
+=== Content-Type Requirements
 
-deprecated[5.3.0, Provide the proper Content-Type header]
+The type of the content sent in a request body must be specified using
+the `Content-Type` header. The value of this header must map to one of
+the supported formats (JSON, YAML, CBOR, SMILE, and NDJSON).
```
Comment (javanna) [inline docs/reference/api-conventions.asciidoc]: we do throw error when one sets CBOR or YAML using _bulk, or NDJSON using e.g. search ?

File: docs/reference/api-conventions.asciidoc [hunk 5]
```diff
@@ -675,22 +675,15 @@ should also be passed with a media type value that indicates the format
...
-A strict mode can be enabled which disables auto-detection and requires
-that all requests with a body have a Content-Type header that maps to
-a supported format. To enabled this strict mode, add the following
-setting to the `elasticsearch.yml` file:
-
-    http.content_type.required: true
-
-The default value is `false`.
+Additionally, when using the `source` query string parameter the
+content type should be specified using the `source_content_type` query
```
Comment (javanna) [inline docs/reference/api-conventions.asciidoc]: must be ? :)

File: docs/reference/migration/migrate_6_0/rest.asciidoc [hunk 6]
```diff
@@ -19,6 +19,9 @@ In previous versions of Elasticsearch, having a proper Content-Type for the data
 Elasticsearch 6.0.0 enforces that all requests with a body must have a supported Content-Type and this type will
 be used when parsing the data.
 
+When using the `source` query string parameter, the `source_content_type` parameter must also be specified with
+the media type of the source.
+
```
Comment (javanna) [inline docs/reference/migration/migrate_6_0/rest.asciidoc]: shall we add a bullet point on the removal of all those setters from the Java api? If we don't want to list them all, shall we just explain the change which type of methods it affects and how to fix the compile error?

**Review by javanna — APPROVED** [review]
left a few comments, LGTM besides those

File: buildSrc/src/main/resources/forbidden/es-test-signatures.txt (hunk 1 above)
Comment (jaymode) [inline buildSrc/src/main/resources/forbidden/es-test-signatures.txt]: ++ I should add all of those here. The difference is the http-signatures file is used only for the rest projects as they do not use the es-test-signatures. 

I could move the http signatures to the buildSrc and then have the default forbidden apis for tests include that file as well. This way we only need to update the file in one place.

File: core/src/main/java/org/elasticsearch/http/HttpTransportSettings.java (hunk 3 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/http/HttpTransportSettings.java]: I can add this as a setting for a node in the rolling-upgrade tests?

File: buildSrc/src/main/resources/forbidden/es-test-signatures.txt (hunk 1 above)
Comment (javanna) [inline buildSrc/src/main/resources/forbidden/es-test-signatures.txt]: I think I would do that, unless there are downsides that I am not seeing

File: core/src/main/java/org/elasticsearch/http/HttpTransportSettings.java (hunk 3 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/http/HttpTransportSettings.java]: sounds good.

**Comment by jaymode** [comment]
Thanks @javanna !

**Pull request merged**